strawberryfields/backends/fockbackend/ops.py

```diff
@@ -318,6 +338,157 @@ def apply_gate_einsum(mat, state, pure, modes, n, trunc):
+        trunc (int): The Hilbert space truncation/cutoff
+        gate (str): The gate that is being applied. This argument determines the selection rules that
+            are used. Options are ``"BSgate"`` and ``"S2gate"``.
+
+    Returns:
+        array[complex]: state after application of the two-mode operation
+    """
+    if pure:
+        t1 = modes[0]
+        t2 = modes[1]
+
+        # moves the modes on which the gate should be applied to the
+        # front indices of the state, for mode 1 and 2 respectively
+        # e.g. from [i1, i2, i3] --> [i2, i3, i1] if the gate is applied to modes 2 and 3
+        switch_list_1 = np.arange(n)
+        switch_list_2 = np.arange(n)
+        switch_list_1[[0, t1]] = switch_list_1[[t1, 0]]
+        switch_list_2[[1, t2]] = switch_list_2[[t2, 1]]
+
+        state = state.transpose(switch_list_1)
+        state = state.transpose(switch_list_2)
+
+        if gate == "BSgate":
+            state = _apply_BS(mat, state, trunc)
+        elif gate == "S2gate":
+            state = _apply_S2(mat, state, trunc)
+        else:
+            raise NotImplementedError("Currently, selection rules are only implemented for the BSgate "
+                                      "and the S2gate. The {} gate is not supported".format(gate))
+
+        state = state.transpose(switch_list_2)
+        ret = state.transpose(switch_list_1)
+    else:
+        t1 = 2 * modes[0]
+        t2 = 2 * modes[1]
+
+        # moves the modes on which the gate should be applied to the
+        # front indices of the state, for mode 1 and 2 respectively
+        # e.g. from [i1, j1, i2, j2, i3, j3] --> [i2, j2, i3, j3, i1, j1]
+        # if the gate is applied to modes 2 and 3
+        switch_list_1 = np.arange(2 * n)
+        switch_list_2 = np.arange(2 * n)
+        switch_list_1[[0, 1, t1, t1+1]] = switch_list_1[[t1, t1+1, 0, 1]]
+        switch_list_2[[0, 1, t2, t2+1]] = switch_list_2[[t2, t2+1, 0, 1]]
+
+        # puts the modes on which the gate should be applied together
+        # e.g. from [i2, j2, i3, j3, i1, j1] --> [i2, i3, j2, j3, i1, j1]
+        # if the gate is applied to modes 2 and 3 (which are already switched to the first two modes)
+        transpose_list = np.arange(2 * n)
+        transpose_list[[t1+1, t2]] = transpose_list[[t2, t1+1]]
+
+        state = state.transpose(transpose_list)
+        state = state.transpose(switch_list_1)
+
+        if gate == "BSgate":
+            state = _apply_BS(mat, state, trunc)
+            state = state.transpose(switch_list_1)
+
+            state = state.transpose(switch_list_2)
+            state = _apply_BS(mat.conj(), state, trunc)
+
+        elif gate == "S2gate":
+            state = _apply_S2(mat, state, trunc)
+            state = state.transpose(switch_list_1)
+
+            state = state.transpose(switch_list_2)
+            state = _apply_S2(mat.conj(), state, trunc)
+
+        else:
+            raise NotImplementedError
```

Comment:
I think one of my previous suggested changes was missed here (see https://github.com/XanaduAI/strawberryfields/pull/292#discussion_r378606418)